### PR TITLE
Check if command exists when creating a systemd timer

### PR DESCRIPTION
**Pull request by jelly**
image

### CODE REVIEW

**Comment by garrett** [comment]
What improvements are there in this PR? All I see is checking if a command exists. Is there more?

Other things that would be nice:

- Editing timers (probably remove and re-add with adjustments)
- Sorting the manually added (probably Cockpit-added timers, really) at the top
- Timer deletion (for Cockpit-added timers)

...but all probably as follow-ups. 

Related: 

- https://github.com/cockpit-project/cockpit/issues/9467

I'll add the list over on that issue.

**Comment by jelly** [comment]
> What improvements are there in this PR? All I see is checking if a command exists. Is there more?
> 
> Other things that would be nice:
> 
> * Editing timers (probably remove and re-add with adjustments)
> * Sorting the manually added (probably Cockpit-added timers, really) at the top
> * Timer deletion (for Cockpit-added timers)
> 
> ...but all probably as follow-ups.
> 
> Related:
> 
> * [services: No ability to edit timers #9467](https://github.com/cockpit-project/cockpit/issues/9467)
> 
> I'll add the list over on that issue.

For now it's just checking if the command is available, I have a local branch for deletion which I plan to create a new PR for.

**Comment by garrett** [comment]
> For now it's just checking if the command is available, I have a local branch for deletion which I plan to create a new PR for.

Minor nitpick. Can we rename the PR's title to help explain what the PR is for? :grin:

**Comment by jelly** [comment]
> > For now it's just checking if the command is available, I have a local branch for deletion which I plan to create a new PR for.
> 
> Minor nitpick. Can we rename the PR's title to help explain what the PR is for? grin

Sure!

**Review by garrett — CHANGES_REQUESTED** [review]
Looks great!

However: Shouldn't the save button go disabled until the entry is changed?

**Comment by jelly** [comment]
> Looks great!
> 
> However: Shouldn't the save button go disabled until the entry is changed?

Good point, shouldn't that apply for all? Read https://www.patternfly.org/v4/components/form/design-guidelines#buttons so I'll see what we can do here.

**Comment by garrett** [comment]
> Good point, shouldn't that apply for all?

We already do this elsewhere in Cockpit. (For example: in Cockpit-Machines create VM, which is where I checked when suggesting this.) Some places _probably_ don't (as software written over time by multiple people can have differences), but should.

I'm suggesting resetting the disabled button state when the form is changed, which is literally "form submit buttons should be disabled until a user starts filling out the form or edits the form in any way", but after a submission.

**Comment by jelly** [comment]
> 



> Looks great!
> 
> However: Shouldn't the save button go disabled until the entry is changed?

Done, please re-check

**Comment by garrett** [comment]
> However: Shouldn't the save button go disabled until the entry is changed?

It doesn't seem to have worked?

image

**Comment by jelly** [comment]
> > > However: Shouldn't the save button go disabled until the entry is changed?
> > 
> > 
> > Done, please re-check
> 
> It doesn't seem to have worked?
> 
> image

Hmmm

image

File: pkg/systemd/services/timer-dialog.jsx [hunk 1]
```diff
@@ -109,9 +110,19 @@ const CreateTimerDialogBody = ({ setIsOpen, owner }) => {
             return false;
 
         setInProgress(true);
-        create_timer({ name, description, command, delay, delayUnit, delayNumber, repeat, specificTime, repeatPatterns, owner })
-                .then(() => setIsOpen(false), exc => {
-                    setDialogError(exc.message);
+
+        // Verify if the command exists
+        const command_parts = command.split(" ");
+        cockpit.spawn(["test", "-f", command_parts[0]], { err: "ignore" })
```
Comment (marusak) [inline pkg/systemd/services/timer-dialog.jsx:116]: Hmm, this assumes absolute path, right?

Comment (jelly) [reply pkg/systemd/services/timer-dialog.jsx:116]: Yes, but I don't think ExecStart supports an relative paths? Reading the man page

https://www.man7.org/linux/man-pages/man5/systemd.service.5.html at ExecStart=
```
         For each of the specified commands, the first argument must
           be either an absolute path to an executable or a simple file
           name without any slashes.
           ```
But as far as I know it does not work with an relative path?

```
[jelle@t14s][/usr/bin]%sudo systemd-run -t ExecStart="uname"
Failed to find executable ExecStart=uname: No such file or directory
```

**Review by marusak — APPROVED** [review]
Thank you!

File: pkg/systemd/services/timer-dialog.jsx (hunk 1 above)
Comment (marusak) [reply pkg/systemd/services/timer-dialog.jsx:116]: Ah right, good to know!

**Pull request merged**